### PR TITLE
with http recorded: Preserve the fulfilment value of the assertion being delegated to

### PR DESCRIPTION
**Pull request by papandreou**
Also, add support for 'with extra info' flag to be used in the test suite.

### CODE REVIEW

File: test/unexpectedMitm.js [hunk 1]
```diff
@@ -1335,6 +1336,13 @@ describe('unexpectedMitm', function () {
             }, 'to yield response', 405);
         });
 
+        it('should preserve the fulfilment value', function () {
+            return expect('foo', 'with http recorded', 'to match', /^(f)o/)
+                .then(function (matches) {
+                    expect(matches, 'to satisfy', {0: 'fo', 1: 'f', index: 0});
+                });
+        });
```
Comment (alexjeffburke) [inline test/unexpectedMitm.js]: This is super pedantic, but to be consistent with other parts of mitm code/tests could we put .then() on the line above and unindent the expect on 1342?

File: test/unexpectedMitm.js [hunk 2]
```diff
@@ -35,11 +35,12 @@ describe('unexpectedMitm', function () {
...
         .addAssertion('<any> with expected http recording <object> <assertion>', function (expect, subject, expectedRecordedExchanges) { // ...
             expect.errorMode = 'nested';
-            expect.args.splice(1, 0, 'with http recorded');
+            expect.args.splice(1, 0, 'with http recorded with extra info');
             return expect.promise(function () {
                 return expect.shift();
-            }).then(function (recordedExchanges) {
+            }).spread(function (value, recordedExchanges) {
                 expect(recordedExchanges, 'to equal', expectedRecordedExchanges);
+                return value;
```
Comment (alexjeffburke) [inline test/unexpectedMitm.js]: Not sure this is necessary in the test suite - we don't chain those assertions.

File: test/unexpectedMitm.js (hunk 1 above)
Comment (papandreou) [inline test/unexpectedMitm.js]: Absolutely, fixed and force-pushed.

File: lib/unexpectedMitm.js [hunk 3]
```diff
@@ -721,14 +721,18 @@ module.exports = {
...
                         if (injectionCallsite) {
                             recordPendingInjection(injectionCallsite, recordedExchanges);
                         }
                     }
-                    return recordedExchanges;
+                    if (expect.flags['with extra info']) {
+                        return [value, recordedExchanges];
+                    } else {
+                        return value;
+                    }
```
Comment (alexjeffburke) [inline lib/unexpectedMitm.js]: This means 'with extra info' doesn't quite return the same order of stuff as for 'with http mocked out' - that's probably totally fine given it's a totally separate assertion :)

Comment (papandreou) [inline lib/unexpectedMitm.js]: You're right. Seemed better than to invent another `with something-something` flag here :)

**Pull request merged**